rancher/rancher
# [release/v2.9] Add new annotation support for RKECluster to allow for cleanup of CAPI machine objects if their corresponding node is not found for a specified duration

## Issue:
https://github.com/rancher/rancher/issues/43686
 
## Problem
In previous versions of Rancher, there was unintended behavior which led to the deletion of Custom Machine objects if their corresponding node objects were no longer present in the downstream cluster. This led to potential data loss, as it could mistakenly remove nodes during etcd restore operations or other disaster recovery scenarios. When we removed this inadvertent behavior, issues were opened pointing out the change in behavior.

## Solution
This PR introduces a new annotation `rke.cattle.io/delete-missing-custom-machines-after` that can be set on either the `clusters.provisioning.cattle.io/v1` or `rkecluster.rke.cattle.io` objects, which will drive logic within the unmanaged machine controller that will delete missing custom machines after a specified duration.
 
## Testing
## Engineering Testing
### Manual Testing
I tested this with Rancher v2.9, with K3s and RKE2. I ran through multiple scenarios, including etcd restoration scenarios.

### Automated Testing
* Test types added/modified:
    * None
* If "None" - Reason: Adding additional tests to the v2prov integration framework is unlikely to succeed due to flakiness.

## QA Testing Considerations
This behavior change is considered opt in. We should ensure that custom machines are not deleted from a cluster if the downstream nodes are deleted if this annotation is not set. We should also ensure that this will not lead to inadvertent node deletion during etcd node restoration.

### Regressions Considerations
As this feature is opt-in, the only regression consideration should be that it does not function if it is not turned on.
